out_stackdriver: support gzip compression

CatherineF-dev opened the pull request:
<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

CatherineF-dev commented:
- Configuration example:
```
     [OUTPUT]
         # https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver
         Name                      stackdriver
         compress_gzip             1
```

- Debug output:
Compression rate is about 95%, from 44562525 bytes to \2141049 bytes.

edsiper commented:
quick review:

- rename the new configuration option to `compress` and support a string value. All output plugins that supports compression uses the same configuration property name, so we don't name it differently, as an example take a look at the Splunk plugin: https://github.com/fluent/fluent-bit/blob/master/plugins/out_splunk/splunk.c#L734-L738 

- sign-off your commits (DCO error)

once those are fixed we can do another review

CatherineF-dev commented:
cc @qingling128

CatherineF-dev commented:
It seems splunk is using `ctx->compress_gzip`.

https://github.com/fluent/fluent-bit/blob/master/plugins/out_splunk/splunk.c#L600

CatherineF-dev commented:
Hi team, have changed to `compress`

CatherineF-dev commented:
/retest

patrick-stephens commented:
Can you link the docs PR?

CatherineF-dev commented:
### Context
#### Objective
Reduce network traffic when sending logs from k8s cluster into Cloud Logging.
It reduced bytes of logs by 95% after using gzip, from 100 bytes  to 5 bytes.
#### Background
Fluentbit stackriver_output is using `Content-Type: application/json` without compression.

### Design
#### Stackdriver output plugin API:
```
[OUTPUT]
         #[ https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver](https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver)
         Name                      stackdriver
         ...
         compress                  "gzip"
```

#### The HTTP header
```
Content-Type: application/json
Content-Encoding: gzip
```

cc @patrick-stephens This is the design doc.

patrick-stephens commented:
@CatherineF-dev apologies, I mean the PR to provide the update to the Fluent Bit documentation. Typically we link the two together so when this PR is merged then released we can merge and release the docs for people using it.

You have added a new configuration parameter `compress` so this needs to be documented here:  https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver#configuration-parameters

To do that, update the Markdown here and submit a PR: https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/stackdriver.md

CatherineF-dev commented:
@patrick-stephens Done, thx!

https://github.com/fluent/fluent-bit-docs/pull/1097

leonardo-albertovich commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2498,14 +2502,26 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
         flb_plg_error(ctx->ins, "cannot retrieve oauth2 token");
         flb_upstream_conn_release(u_conn);
-        flb_sds_destroy(payload_buf);
+        flb_sds_destroy(out_payload_buf);
```
Please revert the changes related to `payload_buf`, add a third variable to hold the final payload (and one for its sice) that's conditionally set to either a copy the raw result of `stackdriver_format` or the output buffer `flb_gzip_compress` produces to make the usage transparent while retaining a proper cleanup procedure.

leonardo-albertovich commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2498,14 +2502,26 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
+    out_payload_buf = (flb_sds_t) payload_buf;
+    out_payload_size = flb_sds_len(payload_buf);
+    if (ctx->compress != NULL && strcmp(ctx->compress, "gzip") == 0) {
+        ret = flb_gzip_compress((void *) payload_buf, flb_sds_len(payload_buf),
```
[flb_gzip_compress](https://github.com/fluent/fluent-bit/blob/ba49c5baa998f5187c6224ebcaf400de9f98370a/src/flb_gzip.c#L85) returns a buffer allocated with `flb_malloc` which is not compatible with `flb_sds_destroy`

CatherineF-dev replied:
Thanks! Done.

leonardo-albertovich commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2596,8 +2616,11 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
     /* Cleanup */
-    flb_sds_destroy(payload_buf);
+    if (out_payload_buf != payload_buf) {
+        flb_free(out_payload_buf);
```
This is related to what I previously mentioned and to make things clearer it'd be better if you renamed the variable to something like `compressed_payload_buffer` which would clearly convey its purpose.

CatherineF-dev replied:
Thanks! Done.

leonardo-albertovich commented:
The splunk used `compress_gzip` as the field name in the context structure so it only compares the string once when the plugin is initialized. That's somewhat understandable under the presumption that other algorithms won't ever be used, however, a better approach would've been defining a set of constants for different supported  algorithms (even if  that meant just defining one option for gzip and one for none), defining an integer field to store the interpreted value and using that in the flush function.

I think this will probably be streamlined in the near future. At the moment we only support gzip and snappy but we will add more algorithms soon (zstd being the first one that comes to mind).

leonardo-albertovich commented:
Ideally, when we do that we should all of the code related to compression to `flb_http_client` so it's transparent to the plugin and we break this code duplication cycle.

JeffLuoo commented:
Hi @leonardo-albertovich, what's the plan for decoupling the compression logic to the fluent-bit client? If that's not near future, I think we still need to have the compression in the out_stackdriver for now.

CatherineF-dev commented:
/hold

Plan to add a safe check for NULL compressed
```
    if (compressed == FLB_TRUE) {
        flb_free(compressed_payload_buffer);
    }
```

CatherineF-dev commented:
@leonardo-albertovich Hi, could you approve again? Thanks!

I rebased and added a safe check for `flb_free(compressed_payload_buffer)`.

Because tests failed before, try rebasing to see whether these failures will be fixed or not.

CatherineF-dev commented:
Hi @JeffLuoo, have you encountered similar test failures before?

I haven't found useful error information.

CatherineF-dev commented:
Hi team, how to merge this PR?

I saw other PRs are merge manually.

CatherineF-dev commented:
cc @qingling128 Need a LGTM on gzip for out_stackdriver plugin, thanks!

leonardo-albertovich commented:
I have already approved this PR, @edsiper will merge it when he's able to.

CatherineF-dev commented:
Ping ~

igorpeshansky commented:
> Plan to add a safe check for NULL compressed
> 
> ```
>     if (compressed == FLB_TRUE) {
>         flb_free(compressed_payload_buffer);
>     }
> ```

BTW, `flb_free` uses `free` [1], which should already handle `NULL` properly [2].

[1] https://github.com/fluent/fluent-bit/blob/master/include/fluent-bit/flb_mem.h#L127 \
[2] https://linux.die.net/man/3/free

CatherineF-dev commented on plugins/out_stackdriver/stackdriver.h:
```diff
@@ -159,6 +159,11 @@ struct flb_stackdriver {
+    /* Payload compression */
+    flb_sds_t compress;
+    /* Internal variable to reduce string comparisons */
+    int compress_gzip;
```
Later refactoring.

CatherineF-dev commented on plugins/out_stackdriver/stackdriver_conf.c:
```diff
@@ -552,6 +552,12 @@ struct flb_stackdriver *flb_stackdriver_conf_create(struct flb_output_instance *
                                                         "Total number of retried records.",
                                                         2, (char *[]) {"status", "name"});
+    /* Compress (gzip) */
+    tmp = flb_output_get_property("compress", ins);
```
Good point. Updated to `FLB_FALSE`. 
Thx!

CatherineF-dev commented on plugins/out_stackdriver/stackdriver_conf.c:
```diff
@@ -552,6 +552,12 @@ struct flb_stackdriver *flb_stackdriver_conf_create(struct flb_output_instance *
                                                         "retried_records_total",
                                                         "Total number of retried records.",
                                                         2, (char *[]) {"status", "name"});
+    /* Compress (gzip) */
+    tmp = flb_output_get_property("compress", ins);
+    ctx->compress_gzip = FLB_FALSE;
+    if (tmp && strcasecmp(tmp, "gzip") == 0) {
+        ctx->compress_gzip = FLB_TRUE;
+    }
```
Got it, thx!

CatherineF-dev commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2520,14 +2523,33 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
+        ret = flb_gzip_compress((void *) payload_buf, flb_sds_len(payload_buf),
+                                &compressed_payload_buffer, &compressed_payload_size);
+        if (ret == -1) {
+            flb_error("[out_stackdriver] cannot gzip payload, disabling compression");
```
Done.

CatherineF-dev commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2520,14 +2523,33 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
+    }
 
     /* Get or renew Token */
     token = get_google_token(ctx);
```
A great point. Done.

CatherineF-dev commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2520,14 +2523,33 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
     }
-    payload_size = flb_sds_len(payload_buf);
+
+    compressed_payload_buffer = (flb_sds_t) payload_buf;
```
Good point. Updated to `compressed_payload_buffer = payload_buf;`

CatherineF-dev commented:
Thanks, have updated! @igorpeshansky

Okay, we can remove this safe check for `flb_free` in future PR.
https://github.com/fluent/fluent-bit/blob/6621b4bbeb2005f928e7ae87114bfe8bdc505def/plugins/out_stackdriver/stackdriver.c#L62-L64

igorpeshansky commented on plugins/out_stackdriver/stackdriver_conf.c:
```diff
@@ -552,6 +552,12 @@ struct flb_stackdriver *flb_stackdriver_conf_create(struct flb_output_instance *
                                                         "Total number of retried records.",
                                                         2, (char *[]) {"status", "name"});
+    /* Compress (gzip) */
+    tmp = flb_output_get_property("compress", ins);
```
Couldn't find the "Unresolve conversation" button, so adding this here to flag that there was a follow-up comment in the earlier (resolved) thread.

igorpeshansky commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2567,13 +2567,14 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
-    flb_http_add_header(c, "Authorization", 13, token, flb_sds_len(token));
     /* Content Encoding: gzip */
     if (compressed == FLB_TRUE) {
+        flb_http_add_header(c, "Content-Type", 12, "text/plain", 10);
```
The `Content-Type: text/plain` on that page is just an example. In this case, the content type is `application/json`, so that's what you should set it to (with `Content-Encoding: gzip`). Setting it to `text/plain` is wrong. Let's instead figure out why `application/json` didn't work.

CatherineF-dev replied:
If it's wrong, what will happen? It won't be json in cloud logging?

I can check cloud logging UI.

CatherineF-dev replied:
Verified, json is working

igorpeshansky commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -2737,6 +2763,11 @@ static struct flb_config_map config_map[] = {
       "Set the resource task id"
     },
+    {
+     FLB_CONFIG_MAP_STR, "compress", NULL,
+     0, FLB_FALSE, 0,
```
I think your indentation is off…
```suggestion
      FLB_CONFIG_MAP_STR, "compress", NULL,
      0, FLB_FALSE, 0,
```

igorpeshansky commented on plugins/out_stackdriver/stackdriver.h:
```diff
@@ -159,6 +159,11 @@ struct flb_stackdriver {
     flb_sds_t job;
     flb_sds_t task_id;
 
+    /* Payload compression */
+    flb_sds_t compress;
```
You can now remove this field…
```suggestion
```

igorpeshansky commented:
cc @leonardo-albertovich as well.
